source/plugin/data/datamanipulators.rst

```diff
@@ -0,0 +1,111 @@
+so much easier' and be right - for getting and setting single values. But the true merit of a data manipulator is
+that it contains *all* data pertaining to a certain component. Let us take a look at another example.
+
+**Code Example: Swapping two data holders' health**
+
+.. code-block:: java
+
+    public void swapHealth(DataHolder targetA, DataHolder targetB) {
+        if (targetA.supports(HealthData.class) && targetB.supports(HealthData.class)) {
+            HealthData healthA = targetA.get(HealthData.class).get();
+            HealthData healthB = targetB.get(HealthData.class).get();
+            targetA.offer(healthB);
+            targetB.offer(healthA);
+        }
+    }
+
+First we check if both targets support HealthData. If they do, we save the health of both in one variable each. We
+don't need to bother with ``Optional`` this time since we verified that ``HealthData`` is supported and the
+``getOrCreate()`` method ensures that even if no data is present, default values are generated.
+
+Then we just offer the saved health data to the *other* target, thus switching their health status with each other.
+
+This example done with ``Keys`` would be a bit longer and more complicated since we'd have to take care of each
+individual key by ourself. And if instead of health we swapped another data manipulator containing even more data
+(maybe ``InvisibilityData`` which even contains a list), we'd have a lot more work to do. But since the data
+holder itself takes care of all data pertaining to it, we could even modify the above function to swap arbitrary
+data between two holders.
+
+**Code Example: Swapping any data manipulator**
+
+.. code-block:: java
+
+    public  <T extends DataManipulator<?,?>> void swapData(DataHolder targetA, DataHolder targetB, Class<T> dataClass) {
+       if (targetA.supports(dataClass) && targetB.supports(dataClass)) {
+           T dataA = targetA.getOrCreate(dataClass).get();
+           T dataB = targetB.getOrCreate(dataClass).get();
+           targetA.offer(dataB);
+           targetB.offer(dataA);
+       }
+    }
+
+The ability to write a function that can just swap any data on a data holder with the same data on another data
+holder demonstrates the core design goal of the Data API: Maximum compatibility across the API.
+
+Mutable vs. Immutable Data Manipulators
+=======================================
+
+To every data manipulator, there is a matching ``ImmutableDataHolder``. For instance, both ``HealthData`` and
+``ImmutableHealthData`` contain the same data, only the latter does not provide any means to make alterations to
+the data.
+
+Conversion between mutable and immutable data manipulators is done via the ``asImmutable()`` and ``asMutable()``
+methods, which each will return a copy of the data. Since the only way to obtain an immutable data manipulator
+from a data holder is obtaining a mutable one and then using ``asImmutable()``, in terms of processing power it
```

Comment:
You don't really need to mention processing power here. Object creation/lookups is really quite cheap in Java.